src/cpp/client/channel_cc.cc

```diff
@@ -71,65 +73,79 @@ class ChannelConnectivityWatcher {
     }
   }
 
-  ~ChannelConnectivityWatcher() {
-    cq_.Shutdown();
-    if (thd_id_ != 0) {
-      gpr_thd_join(thd_id_);
-    }
-  }
-
   void WatchStateChangeImpl() {
     bool ok = false;
     void* tag = NULL;
     CompletionQueue::NextStatus status = CompletionQueue::GOT_EVENT;
-    while (status != CompletionQueue::SHUTDOWN) {
+    while (true) {
       status = cq_.AsyncNext(&tag, &ok, gpr_inf_past(GPR_CLOCK_REALTIME));
       // Make sure we've seen 2 TIMEOUTs before going to sleep
       if (status == CompletionQueue::TIMEOUT) {
         status = cq_.AsyncNext(&tag, &ok, gpr_inf_past(GPR_CLOCK_REALTIME));
+        if (status == CompletionQueue::TIMEOUT) {
+          gpr_sleep_until(
+              gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
+                           gpr_time_from_millis(kConnectivityCheckIntervalMsec,
+                                                GPR_TIMESPAN)));
+          continue;
+        }
       }
-      if (status == CompletionQueue::TIMEOUT) {
-        gpr_sleep_until(
-            gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
-                         gpr_time_from_millis(kConnectivityCheckIntervalMsec,
-                                              GPR_TIMESPAN)));
-      } else if (status == CompletionQueue::GOT_EVENT) {
-        ChannelState* channel_state = static_cast<ChannelState*>(tag);
-        channel_state->state = grpc_channel_check_connectivity_state(
-            channel_state->channel, false);
-        if (channel_state->state == GRPC_CHANNEL_SHUTDOWN) {
-          void* shutdown_tag = NULL;
-          channel_state->shutdown_cq.Next(&shutdown_tag, &ok);
-          delete channel_state;
-        } else {
-          TagSaver* tag_saver = new TagSaver(channel_state);
-          grpc_channel_watch_connectivity_state(
-              channel_state->channel, channel_state->state,
-              gpr_inf_future(GPR_CLOCK_REALTIME), cq_.cq(), tag_saver);
+      ChannelState* channel_state = static_cast<ChannelState*>(tag);
+      channel_state->state =
+          grpc_channel_check_connectivity_state(channel_state->channel, false);
+      if (channel_state->state == GRPC_CHANNEL_SHUTDOWN) {
+        void* shutdown_tag = NULL;
+        channel_state->shutdown_cq.Next(&shutdown_tag, &ok);
+        delete channel_state;
+        if (gpr_unref(&ref_)) {
+          gpr_mu_lock(&g_watcher_mu_);
+          delete g_watcher_;
+          g_watcher_ = nullptr;
+          gpr_mu_unlock(&g_watcher_mu_);
+          break;
         }
+      } else {
+        TagSaver* tag_saver = new TagSaver(channel_state);
+        grpc_channel_watch_connectivity_state(
+            channel_state->channel, channel_state->state,
+            gpr_inf_future(GPR_CLOCK_REALTIME), cq_.cq(), tag_saver);
       }
     }
   }
 
-  void StartWatching(grpc_channel* channel) {
+  static void StartWatching(grpc_channel* channel) {
+    gpr_once_init(&g_connectivity_watcher_once_, InitConnectivityWatcherOnce);
```

Comment:
I assume that `gpr_once_init()` is not happy directly calling a static member function for some reason?

Reply:
`gpr_once_init()` is not willing to use the static member function `ChannelConnectivityWatcher::InitOnce()`, so I had to add a wrapper `InitConnectivityWatcherOnce()`.